src/box/memtx_engine.cc

```diff
@@ -320,8 +384,15 @@ memtx_space_on_replace(struct space *space)
 
 	/* Start building PK of the next space to recover if required. */
 	struct memtx_space *memtx_space = (struct memtx_space *)space;
-	if (memtx_space->replace == memtx_space_replace_build_next)
+	if (memtx_space->replace == memtx_space_replace_build_next) {
 		index_begin_build(space->index[0]);
+		/* Use the sort data for secondary keys if required. */
+		if (memtx->recovery.sort_data_reader != NULL) {
+			memtx_sort_data_reader_space_init(
+				memtx->recovery.sort_data_reader,
+				space->def->id);
+		}
+	}
```

Comment:
What I dislike about recovery in this patch is that we now have two independent ways of secondary index recovery:
1. Through `index_begin_build`, `index_build_next`, `index_end_build` callbacks.
2. Using the sort data reader.

Come to think of it, the two ways aren't that different:
- For both of them we accumulate all tuples from the primary index. We just use different containers (array for sort; hash table for sort data).
- Then we build the secondary index using those tuples, either by sorting the accumulated tuples or by restoring the order from the file.

Would it be possible to unify the two ways? Ideally, refactor the index callbacks in such a way that the differentiation happens at the lowest level.